src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.cs

```diff
@@ -0,0 +1,212 @@
+
+        private readonly SafeBCryptKeyHandle _keyHandle;
+        private readonly HashAlgorithmName _hashAlgorithm;
+
+        public void Dispose()
+        {
+            _keyHandle.Dispose();
+        }
+
+        public void DeriveBytes(byte[] label, byte[] context, Span<byte> destination)
+        {
+            DeriveBytes(new ReadOnlySpan<byte>(label), new ReadOnlySpan<byte>(context), destination);
+        }
+
+        public unsafe void DeriveBytes(ReadOnlySpan<byte> label, ReadOnlySpan<byte> context, Span<byte> destination)
+        {
+            if (destination.Length == 0)
+            {
+                return;
+            }
+
+            Debug.Assert(destination.Length <= 0x1FFFFFFF);
+            Debug.Assert(_hashAlgorithm.Name is not null);
+
+            fixed (byte* pLabel = label)
+            fixed (byte* pContext = context)
+            fixed (byte* pDestination = destination)
+            fixed (char* pHashAlgorithm = _hashAlgorithm.Name)
+            {
+                const int BCryptBufferLength = 3;
+                BCryptBuffer* buffers = stackalloc BCryptBuffer[BCryptBufferLength];
+
+                buffers[0].BufferType = CngBufferDescriptors.KDF_LABEL;
+                buffers[0].pvBuffer = (IntPtr)pLabel;
+                buffers[0].cbBuffer = label.Length;
+                buffers[1].BufferType = CngBufferDescriptors.KDF_CONTEXT;
+                buffers[1].pvBuffer = (IntPtr)pContext;
+                buffers[1].cbBuffer = context.Length;
+                buffers[2].BufferType = CngBufferDescriptors.KDF_HASH_ALGORITHM;
+                buffers[2].pvBuffer = (IntPtr)pHashAlgorithm;
+                buffers[2].cbBuffer = (_hashAlgorithm.Name.Length + 1) * 2; // +1 for the null terminator.
+
+                Interop.BCrypt.BCryptBufferDesc bufferDesc;
+                bufferDesc.ulVersion = Interop.BCrypt.BCRYPTBUFFER_VERSION;
+                bufferDesc.cBuffers = BCryptBufferLength;
+                bufferDesc.pBuffers = (IntPtr)buffers;
+
+                NTSTATUS deriveStatus = Interop.BCrypt.BCryptKeyDerivation(
+                    _keyHandle,
+                    &bufferDesc,
+                    pDestination,
+                    destination.Length,
+                    out uint resultLength,
+                    dwFlags: 0);
+
+                if (deriveStatus != NTSTATUS.STATUS_SUCCESS)
+                {
+                    throw Interop.BCrypt.CreateCryptographicException(deriveStatus);
+                }
+
+                if (destination.Length != resultLength)
+                {
+                    Debug.Fail("BCryptKeyDerivation resultLength != destination.Length");
+                    throw new CryptographicException();
+                }
+            }
+        }
+
+        public void DeriveBytes(ReadOnlySpan<char> label, ReadOnlySpan<char> context, Span<byte> destination)
+        {
+            using (Utf8DataEncoding labelData = new Utf8DataEncoding(label, stackalloc byte[128]))
```

Comment:
Consider pulling this out as a const so that it can be updated everywhere by changing a single line. Also consider what a common label \& context length might be. A scratch buffer of 128 bytes means that any source span \>= 43 chars will hit the crypto pool.

Reply:
Labels are mostly driven by RFCs or are usually on the small side. Contexts... can really go anywhere in size. I ended up bumping it to 256 because we should have the stack space for it.